### PR TITLE
[Embedder] Send key data through message channel

### PR DESCRIPTION
**Pull request by dkwingsmt**
This PR changes how embedder API's `SendKeyData` sends `ui.KeyData` to the framework.

Currently the packet is sent over FFI through a series of `DispatchKeyData` methods that drill all the way from `Window` to `EmbedderEngine`. Not only is it cumbersome, but it also bypasses the existing buffering mechanism, leading to a crash as described in https://github.com/flutter/flutter/issues/89748#issuecomment-948567892. 

This PR changes the packet to be sent over the existing platform messenger, reusing the entirety of its code path and functionalities while keeping the embedder API unchanged.

This partially fixes https://github.com/flutter/flutter/issues/89748#issuecomment-948567892. Since the buffer size is only 1, if 2 or more events are sent during the restart (which usually means holding multiple keys) the crash will still occur. There's a way to resize the buffer, but I haven't found a way to actually do it yet, and will explore as the next step.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by Hixie** [comment]
We should document the protocol somewhere so that new embedders know what to implement.

**Comment by chinmaygarde** [comment]
> We should document the protocol somewhere so that new embedders know what to implement.

@Hixie: The public embedder API defines this in the [`FlutterKeyEvent`](https://github.com/flutter/engine/blob/67a16c3b6fbe6ea781801454edf01d9b24b6a2ec/shell/platform/embedder/embedder.h#L683) struct. I believe this was in response to the earlier approach where embedders had to reverse-engineer the payload from what iOS or Android were doing. The protocol in this patch remains unchanged (and is documented). It just affects how the information is ferried to the framework.

> Currently the packet is sent over FFI

Ah, I thought you meant `dart:ffi` for second and I was confused. That term is really overloaded. Yeah, making use of platform of the channels mechanism makes sense.

> while keeping the embedder API unchanged.

Awesome. Yeah, this approach makes sense.

**Comment by dkwingsmt** [comment]
Awesome. Thank you. I'll make some final changes before sending over for reviewing.

File: shell/platform/embedder/embedder.cc [hunk 1]
```diff
@@ -1593,6 +1611,40 @@ static inline flutter::KeyEventType MapKeyEventType(
...
+      sizeof(FlutterPlatformMessage),  // struct_size
+      channel,                         // channel
+      data,                            // message
+      size,                            // message_size
+      response_handle,                 // response_handle
+  };
+
+  result = FlutterEngineSendPlatformMessage(engine, &message);
+  FlutterEngineResult result2 =
+      FlutterPlatformMessageReleaseResponseHandle(engine, response_handle);
```
Comment (dnfield) [inline shell/platform/embedder/embedder.cc:1640]: nit: I think we can move this down below the if right? Also, please use a more descriptive name than `result2`, or, alternatively, just change the `return result2` to `return FlutterPlatofrmMessageReleaseResponseHandle(engine, response_handle);`.

**Comment by dnfield** [comment]
What's the impact on latency here, if any?

I'm curious about why we don't experience the same problem(s) with pointer data packets too.

File: shell/platform/embedder/embedder.cc (hunk 1 above)
Comment (dkwingsmt) [reply shell/platform/embedder/embedder.cc:1640]: No. It's exactly what this awkward structure means: Even if the `SendPlatformMessage` call fails, we still need to release the response handle. I'll add a comment for it.

Comment (dnfield) [reply shell/platform/embedder/embedder.cc:1640]: Ahh ok, it's doing some kind of RAII thing here?

At any rate, it should be named better than `result2` :)

**Comment by dkwingsmt** [comment]
> What's the impact on latency here, if any?

I don't think it impact latency at all, since the old way is basically implementing a parallel platform messenger.

> I'm curious about why we don't experience the same problem(s) with pointer data packets too.

The keyboard system uses a much stricter regularization check, which asserts for any irregular event sequence, while the pointer data converter is much more tolerant and will fill the gap if possible.

**Comment by skia-gold** [comment]
Gold has detected about 69 new digest(s) on patchset 15.
View them at https://flutter-engine-gold.skia.org/cl/github/29638

**Pull request merged**